File: Chapter02/example_topic_consumer.rb
#	Require client library
require "bunny"

#	Read RABBITMQ_URI from ENV
connection = Bunny.new ENV['RABBITMQ_URI']

#	Start a communication session with RabbitMQ
connection.start
channel = connection.create_channel

def taxi_topic_subscribe(channel, taxi, type)
  # Declare a queue for a given taxi
  queue = channel.queue(taxi, durable: true)

  # Declare a topic exchange
  exchange = channel.topic('taxi-topic', durable: true, auto_delete: true)

  #	Bind the queue to the exchange
  queue.bind(exchange, routing_key: type)

  #	Bind the queue to the exchange to make sure the taxi will get any order
  queue.bind(exchange, routing_key: 'taxi')

  #	Subscribe from the queue
  queue.subscribe(block:true,manual_ack: false) do |delivery_info, properties, payload|

    process_order(payload)
  end
end

taxi = "taxi.3"
taxi_topic_subscribe(channel, taxi, "taxi.eco.3")

File: csharp_code/Chapter02/example_topic_consumer.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;
using RabbitMQ.Client;
using RabbitMQ.Client.Events;
using RabbitMQ.Samples.Common;

namespace RabbitMQ.Samples.Chapter02
{
    public class ExampleTopicConsumer
    {
        public static void TaxiTopicSubscribe(RabbitMqConnection rabbitMqConnection)
        {
            var channel = rabbitMqConnection.Channel;

            // Creates a queue named 'taxi.3' with 2 bindings in the topic exchange
            // First binding receives all messages with routing key 'taxi.eco.*'
            // second binding receives all messages with routing key 'taxi'
            TaxiTopicSubscribe(channel, queueName: "taxi.3", "taxi.eco.*");
        }

        private static void TaxiTopicSubscribe(IModel channel, string queueName, string type)
        {
            // Declare a queue for a given taxi
            channel.QueueDeclare(queueName, durable: true);

            // Bind the queue to the exchange
            channel.QueueBind(queue: queueName,
                exchange: Chapter2Constants.TopicExchange,
                routingKey: type);

            // Bind the queue to the exchange to make sure the taxi will get any order
            channel.QueueBind(queue: queueName,
                exchange: Chapter2Constants.TopicExchange,
                routingKey: "taxi");

            Consume(channel, queueName);
        }


        private static void Consume(IModel channel, string queueName)
        {
            // create a consumer
            var consumer = new EventingBasicConsumer(channel);

            consumer.Received += (sender, args) =>
            {
                var body = args.Body.ToArray();

                var message = Encoding.UTF8.GetString(body);

                ProcessOder(message);
            };

            var consumerTag = channel
                .BasicConsume(
                    queue: queueName,
                    autoAck: true, // manual_ack: false
                    consumer: consumer,
                    consumerTag: nameof(ExampleTopicConsumer),
                    noLocal: false,
                    exclusive: false,
                    arguments: null);
        }

        private static void ProcessOder(string message)
        {
            Console.WriteLine("Handling taxi order");
            Console.WriteLine(message);
            Thread.Sleep(500);

            Console.WriteLine("Processing done");
        }
    }
}
